### PR TITLE
Vendor our dispatcher.Message and wireformat.

### LINKED ISSUE #19: Vendor wireformat from sidecar
Related to https://github.com/projectriff/function-sidecar/issues/34

### PR DESCRIPTION
**Pull request by ericbottard**
Harmonize sarama package path

Fixes #19

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -24,3 +24,8 @@
 [[constraint]]
   name = "github.com/satori/go.uuid"
   version = "v1.1.0"
+
+[[constraint]]
+  name = "github.com/projectriff/function-sidecar"
+  branch = "issue_33"
```
Comment (markfisher) [inline Gopkg.toml:30]: this should be removed from this PR (the `function-sidecar` PR will be merged first)

File: cmd/http-gateway.go [hunk 2]
```diff
@@ -78,28 +80,35 @@ func replyHandler(producer sarama.AsyncProducer, replies *repliesMap) http.Handl
...
-		replyChan := make(chan message.Message)
+		replyChan := make(chan dispatcher.Message)
 		replies.put(correlationId, replyChan)
 
-		msg := message.Message{Payload: b, Headers: make(map[string]interface{})}
-		propagateIncomingHeaders(r, &msg)
-		msg.Headers[CorrelationId] = correlationId
+		msg := dispatcher.NewMessage(b, make(map[string][]string))
+		propagateIncomingHeaders(r, msg)
+		log.Printf("Before %v", msg)
```
Comment (markfisher) [inline cmd/http-gateway.go:88]: were these before/after log statements left intentionally? (if so, probably need a bit more context)

File: cmd/http-gateway.go [hunk 3]
```diff
@@ -78,28 +80,35 @@ func replyHandler(producer sarama.AsyncProducer, replies *repliesMap) http.Handl
...
 		if err != nil {
-			log.Printf("Error encoding message: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
 			return
 		}
-		kafkaMsg := &sarama.ProducerMessage{Topic: topic, Value: sarama.ByteEncoder(bytesOut)}
+		kafkaMsg.Topic = topic
+
+		encoded, _ := kafkaMsg.Value.Encode()
+		log.Printf("Kafka: %v", string(encoded))
```
Comment (markfisher) [inline cmd/http-gateway.go:100]: these 2 log statements also left behind from debugging?

File: cmd/http-gateway.go [hunk 4]
```diff
@@ -78,28 +80,35 @@ func replyHandler(producer sarama.AsyncProducer, replies *repliesMap) http.Handl
...
 
 		select {
 		case producer.Input() <- kafkaMsg:
 			select {
 			case reply := <-replyChan:
 				replies.delete(correlationId)
-				p := reply.Payload
-				propagateOutgoingHeaders(&reply, w)
-				w.Write(p.([]byte)) // TODO equivalent of Spring's HttpMessageConverter handling
+				//log.Printf("Got a reply: %v", reply)
```
Comment (markfisher) [inline cmd/http-gateway.go:109]: should remove

**Comment by markfisher** [comment]
should the wireformat and dispatcher code move to an independent lib that is shared between the sidecar and the gateway? (seems odd to have the sidecar as a dependency of the gateway)

File: Gopkg.toml (hunk 1 above)
Comment (ericbottard) [reply Gopkg.toml:30]: Correct. I assume you'll need this to test though. I can remove, as you prefer

File: cmd/http-gateway.go (hunk 2 above)
Comment (ericbottard) [reply cmd/http-gateway.go:88]: All of these are debug leftovers, will remove. They're interesting for the little "quizz", if I remember correctly though

Comment (markfisher) [reply cmd/http-gateway.go:88]: can wait then... will remove on merge

File: Gopkg.toml (hunk 1 above)
Comment (markfisher) [reply Gopkg.toml:30]: can remove on merge

**Comment by ericbottard** [comment]
> should the wireformat and dispatcher code move to an independent lib that is shared between the sidecar and the gateway? (seems odd to have the sidecar as a dependency of the gateway)

Yes, I thought about this. Could do, to be 100% clean, although given the number of projects involved (2) and the size of them, I'm not sure it's needed in practice (means one more vendoring hop and PR coordination, but heh...)

**Comment by markfisher** [comment]
We can hold off on that change for now. I just wanted to throw out the idea. It might be more relevant if we also end up adding an "ambassador" for use by event sources.

**Comment by ericbottard** [comment]
As I work on porting the function controller to go, I realize there could be value in at least one other shareable project between the function sidecar and the function controller:
Currently, the consumer group is the function name. But the true identity of a function is namespace+name, which should somehow be used to derive the consumer group. That logic could benefit from a shared library

File: Gopkg.lock [hunk 5]
```diff
@@ -54,6 +54,13 @@
   revision = "f051bb7f1d1aaf1b5a665d74fb6b0217712c69f7"
   version = "v0.1.1"
 
+[[projects]]
```
Comment (markfisher) [inline Gopkg.lock:57]: need to update this for merging?

Comment (ericbottard) [reply Gopkg.lock:57]: Can be done afterwards

**Pull request merged**